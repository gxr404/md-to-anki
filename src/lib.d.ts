declare module '@gxr1020/anki-apkg-export'
