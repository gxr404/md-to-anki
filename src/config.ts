import { readFileSync } from 'node:fs'
import type { PartialDeep } from 'type-fest'

import { merge } from './utils'

export function getDefaultConfig(): IConfig {
  return {
    code: {
      defaultLanguage: 'bash',
      template: 'dark', // [default | dark]
    },
    card: {
      // 卡片分割: ##.. xxx
      separator: '(?=^#{2,}\\s)',
      // 卡片正反面分割: 正面 % 反面
      frontBackSeparator: '%',
      // 标签: [#标签]
      tagPattern: '^\\[#(.*)\\]',
    },
    deck: {
      titleSeparator: '^#\\s',
      defaultName: 'mdanki',
    },
    template: {
      formats: {
        question: readFileSync(new URL('../resources/question.html', import.meta.url)).toString(),
        answer: readFileSync(new URL('../resources/answer.html', import.meta.url)).toString(),
        css: readFileSync(new URL('../resources/formats.css', import.meta.url)).toString(),
      },
    },
  }
}

export interface IConfig {
  code: {
    defaultLanguage: string
    template: string
  }
  card: {
    separator: string
    frontBackSeparator: string
    tagPattern: string
  }
  deck: {
    titleSeparator: string
    defaultName: string
  }
  template: {
    formats: {
      question: string
      answer: string
      css: string
    }
  }
}

let config: IConfig

export function initConfig(userConfig: PartialDeep<IConfig>): IConfig {
  config = getDefaultConfig()
  config = merge(config, userConfig) as IConfig
  return config
}

export function getConfig() {
  return config
}
