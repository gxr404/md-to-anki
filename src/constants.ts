export const AVAILABLE_FILE_EXTENSIONS = ['.md', '.markdown']

export const THEME = {
  DRACULA: 'dracula',
  MINIMAL: 'minimal',
  NORD: 'nord',
}
