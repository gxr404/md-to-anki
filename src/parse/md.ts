import { Marked } from 'marked'
import { markedHighlight } from 'marked-highlight'
import { codeToHtml } from 'shiki'

let marked: Marked

async function initMarked() {
  if (marked)
    return marked
  // const highlighter = await initHighlighter()
  marked = new Marked({
    gfm: true,
    breaks: true,
    pedantic: false,
    async: true,
    ...markedHighlight({
      langPrefix: 'core lang-',
      async: true,
      async highlight(code, lang) {
        const langAlias: any = {
          sol: 'solidity',
        }
        const htmlStr = await codeToHtml(code, {
          lang: langAlias[lang] || lang,
          theme: 'vitesse-dark',
        })
        return htmlStr
      },
    }),
  })
  return marked
}

export function getMarked() {
  return marked
}

// let highlighter: HighlighterCore

// async function initHighlighter() {
//   if (highlighter)
//     return highlighter
//   // highlighter = await getHighlighterCore({
//   //   themes: [
//   //     import('shiki/themes/vitesse-dark.mjs'),
//   //   ],
//   //   langs: [
//   //     import('shiki/langs/javascript.mjs'),
//   //     import('shiki/langs/css.mjs'),
//   //     import('shiki/langs/bash.mjs'),
//   //     import('shiki/langs/bat.mjs'),
//   //     import('shiki/langs/c.mjs'),
//   //     import('shiki/langs/coffeescript.mjs'),
//   //     import('shiki/langs/go.mjs'),
//   //     import('shiki/langs/json.mjs'),
//   //     import('shiki/langs/graphql.mjs'),
//   //     import('shiki/langs/java.mjs'),
//   //     import('shiki/langs/less.mjs'),
//   //     import('shiki/langs/markdown.mjs'),
//   //     import('shiki/langs/php.mjs'),
//   //     import('shiki/langs/python.mjs'),
//   //     import('shiki/langs/ruby.mjs'),
//   //     import('shiki/langs/rust.mjs'),
//   //     import('shiki/langs/sass.mjs'),
//   //     import('shiki/langs/stylus.mjs'),
//   //     import('shiki/langs/typescript.mjs'),
//   //     import('shiki/langs/vim.mjs'),
//   //     import('shiki/langs/jsx.mjs'),
//   //     import('shiki/langs/vue.mjs'),
//   //     import('shiki/langs/sql.mjs'),
//   //     import('shiki/langs/csharp.mjs'),
//   //     import('shiki/langs/yaml.mjs'),
//   //   ],
//   //   langAlias: {
//   //     'c#': 'csharp',
//   //     'md': 'markdown',
//   //     'rb': 'ruby',
//   //     'py': 'python',
//   //     'sh': 'bash',
//   //   },
//   //   loadWasm: getWasm,
//   // })
//   const { getHighlighter, bundledLanguages } = await import('shiki')
//   highlighter = await getHighlighter({
//     themes: ['vitesse-dark'],
//     langs: Object.keys(bundledLanguages),
//     // [
//     //   'javascript',
//     //   'css',
//     //   'bash',
//     //   'bat',
//     //   'c',
//     //   'coffeescript',
//     //   'go',
//     //   'json',
//     //   'graphql',
//     //   'java',
//     //   'less',
//     //   'markdown',
//     //   'php',
//     //   'python',
//     //   'ruby',
//     //   'rust',
//     //   'sass',
//     //   'stylus',
//     //   'typescript',
//     //   'vim',
//     //   'jsx',
//     //   'vue',
//     //   'sql',
//     //   'csharp',
//     //   'yaml',
//     //   'http',
//     //   'proto',
//     //   'cpp',
//     //   'solidity',
//     // ],
//     langAlias: {
//       'c#': 'csharp',
//       'md': 'markdown',
//       'rb': 'ruby',
//       'py': 'python',
//       'sh': 'bash',
//       'sol': 'solidity',
//       'c++': 'cpp',
//     },
//   })
//   return highlighter
// }

export async function md2Html(mdStr: string) {
  const marked = await initMarked()
  const htmlStr = marked.parse(mdStr)
  return htmlStr
}
