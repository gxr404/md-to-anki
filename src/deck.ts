import { writeFile } from 'node:fs/promises'
import { AnkiExport } from '@gxr1020/anki-apkg-export'
import { getConfig } from './config'
import type { ICard } from './parse'
import type { IMedia } from './parse/media'

interface IExportDeckParams {
  cards: ICard[]
  mediaList: IMedia[]
  targetFile: string
  deckName: string
}

export async function exportDeck(params: IExportDeckParams) {
  const { cards, mediaList, targetFile, deckName } = params
  const config = getConfig()
  const { formats } = config.template
  const ankiExport = await AnkiExport(deckName, {
    css: formats.css,
    answerFormat: formats.answer,
    questionFormat: formats.question,
  })

  // cards
  cards.forEach((card) => {
    const { front, back, tags } = card
    ankiExport.addCard(front, back, { tags })
  })

  // media
  mediaList.forEach((media) => {
    ankiExport.addMedia(media.fileName, media.data)
  })

  const zip = await ankiExport.save()
  await writeFile(targetFile, zip, 'binary')
}
