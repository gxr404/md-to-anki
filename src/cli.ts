import { readFileSync } from 'node:fs'
import process from 'node:process'
import cac from 'cac'
import { THEME } from './constants'
import { main } from './index'

export interface ICliOption {
  target: string
  config: string
  deckName: string
  theme: string
}

function createCli() {
  const { name, version } = JSON.parse(
    readFileSync(new URL('../package.json', import.meta.url)).toString(),
  )
  const cli = cac(name)

  cli
    .command('<mdFile>', 'markdown文件')
    .option('-t, --target <targetFile>', '输出的anki文件名 eg: "-d targe.apkg"')
    .option('-c, --config <configFile>', '配置文件 eg: "-c ./config.json"')
    .option('-s, --theme <theme>', '样式主题可选 nord/minimal/dracula', {
      default: THEME.NORD,
    })
    .option('-d, --deckName <deckName>', '卡片组名 eg: "-d Test", Default: 取md文件中的"# xx"')
    .action(async (mdFile: string, options: ICliOption) => {
      try {
        await main({
          sourcePath: mdFile,
          ...options,
        })
        // await main(mdFile, options)
      }
      catch (err) {
        console.error(err instanceof Error ? err.message : 'unknown exception')
      }
    })

  cli.help()
  cli.version(version)
  return {
    cli,
    parse() {
      try {
        cli.parse()
      }
      catch (err) {
        console.error(err instanceof Error ? err.message : 'unknown exception')
        process.exit(1)
      }
    },
  }
}

const { parse } = createCli()
parse()
