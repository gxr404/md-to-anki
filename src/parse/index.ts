import { readFile } from 'node:fs/promises'
import { getConfig } from '../config'
import { removeMdComments, trimArray } from '../utils'
import { md2Html } from './md'
import { addMediaToCards, mediaList, resetMediaList } from './media'

export async function parseMarkdown(file: string) {
  const mdBuffer = await readFile(file)
  const mdString = removeMdComments(mdBuffer.toString())
  const { rawCards, deckName } = await splitByCards(mdString)
  // console.time('parseCardList')
  const { cards, media } = await parseCardList(rawCards, file)
  // console.timeEnd('parseCardList')

  return {
    cards,
    media,
    deckName,
  }
}

export interface ICard {
  front: string
  back: string
  tags: string[]
}

export async function splitByCards(mdString: string) {
  const config = getConfig()
  const rawCards = mdString
    .split(new RegExp(config.card.separator, 'm'))
    .map((line: string) => line.trim())
  return {
    deckName: getDeckName(rawCards),
    rawCards,
  }
}

function getDeckName(rawCards: string[]) {
  const config = getConfig()
  const titleReg = new RegExp(config.deck.titleSeparator)
  let deckName = rawCards
    .find((str: string) => titleReg.test(str)) ?? ''
  deckName = deckName.split('\n')?.[0] || ''
  if (!deckName)
    return config.deck.defaultName

  return deckName.replace(/(#\s|\n)/g, '').trim()
}

export async function parseCardList(rawCards: string[], sourceFile: string) {
  resetMediaList()
  const dirtyCards = await Promise.all(
    rawCards.map(async (str: string) => {
      // console.time(`parseCard#${i}`)
      const card = await parseCard(str)
      // console.timeEnd(`parseCard#${i}`)

      if (card)
        await addMediaToCards([card], sourceFile)
      return card
    }),
  )

  const cards = dirtyCards
    .filter(card => Boolean(card))
    // card should have front and back sides
    .filter(card => card?.front && card?.back) as ICard[]
  return {
    cards,
    media: mediaList,
  }
}

export async function parseCard(cardStr: string = '') {
  const config = getConfig()
  const splitCardReg = new RegExp(`^${config.card.frontBackSeparator}$`, 'm')
  // 卡片的内容 按每一行分割
  const cardLines = cardStr
    .split(splitCardReg)
    .map((item: string) => item.split('\n'))
    .map((arr: string[]) => arr.map((str: string) => str.trimEnd()))

  // 不允许卡片只有正面(not allowed cards with only front side)
  if (cardLines.length === 1 && !cardLines[0].filter(Boolean).length)
    return null

  const { front, back, tags } = parseCardLines(cardLines)

  // 一般都是转html不存在不转的情况
  // if (!this.options.convertToHtml) {
  //   return new Card(front.join(), back.join(), tags);
  // }
  const frontHtml = await md2Html(front.join('\n'))
  const backHtml = await md2Html(back.join('\n'))

  return {
    front: frontHtml,
    back: backHtml,
    tags,
  }
}

function parseTags(line: string) {
  const config = getConfig()
  const tagReg = new RegExp(config.card.tagPattern)
  const data = line.split(' ')
    .map((str: string) => str.trim())
    .map((str) => {
      const parts = tagReg.exec(str)
      return parts?.[1] ?? ''
    })
    .filter(Boolean)

  return data
}

function parseCardLines(cardLines: string[][]) {
  const config = getConfig()
  const tagReg = new RegExp(config.card.tagPattern)

  const front: string[] = []
  const back: string[] = []
  const tags: string[] = []

  function fillTagAndBack(line: string) {
    if (tagReg.test(line) && line) {
      tags.push(...parseTags(line))
      return
    }
    // back 第一个元素如果是空 则跳过
    if (back.length === 0 && !line)
      return
    back.push(line)
  }

  // 没设置卡片分割符 默认 正面为第一行，其余为反面
  if (cardLines.length === 1) {
    trimArray(cardLines[0])
      .forEach((line, index) => {
        // 默认 正面为第一行
        if (index === 0) {
          front.push(line)
          return
        }
        fillTagAndBack(line)
      })
  }
  else {
    front.push(...cardLines[0])
    trimArray(cardLines[1])
      .forEach((line: string) => {
        fillTagAndBack(line)
      })
  }

  return {
    front: trimArray(front),
    back: trimArray(back),
    tags: trimArray(tags),
  }
}

export * from './media'
