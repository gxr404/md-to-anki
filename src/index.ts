import path from 'node:path'
import process from 'node:process'
import { readFile } from 'node:fs/promises'
// import ora from 'ora'

import type { ICliOption } from './cli'
import { validatePath } from './utils'
import { AVAILABLE_FILE_EXTENSIONS } from './constants'
import { transforms } from './transforms'
import { initConfig } from './config'
import type { IConfig } from './config'
import { exportDeck } from './deck'

interface IOptions extends ICliOption {
  sourcePath: string
}

export async function main(options: IOptions) {
  const { sourcePath: rawSourcePath, deckName, target, config: rawUserConfigPath } = options
  const sourcePath = path.resolve(rawSourcePath)
  const userConfigPath = rawUserConfigPath ? path.resolve(rawUserConfigPath) : ''
  const validate = validatePath(sourcePath, AVAILABLE_FILE_EXTENSIONS)

  if (!validate)
    process.exit(1)
  let userConfig: Partial<IConfig> = {}

  if (userConfigPath) {
    const configValidate = validatePath(userConfigPath, ['.json'])
    if (!configValidate)
      process.exit(1)
    const userConfigBuffer = await readFile(userConfigPath)
    userConfig = JSON.parse(userConfigBuffer.toString())
  }
  let targetFile
  if (!target) {
    const ext = path.extname(sourcePath)
    targetFile = ext ? sourcePath.replace(ext, '.apkg') : `${sourcePath}.apkg`
  }
  else {
    targetFile = path.resolve(target)
  }
  const config = initConfig(userConfig)

  const ora = await import('ora')
  const spinner = ora.default('Transforms...').start()
  // console.time('transforms')
  const {
    cards = [],
    media,
    parseDeckName = config.deck.defaultName,
  } = await transforms({
    sourcePath,
    deckName: deckName || '',
    targetFile: target || '',
  })
  // console.timeEnd('transforms')
  if (!cards.length) {
    spinner.fail('No cards found. Check you markdown file(s)')
    // console.error('No cards found. Check you markdown file(s)')
    process.exit(1)
  }
  const realDeckName = deckName || parseDeckName || config.deck.defaultName

  spinner.text = 'Export...'

  try {
    await exportDeck({
      cards,
      mediaList: media,
      targetFile,
      deckName: realDeckName,
    })
  }
  catch (e: any) {
    console.error(e?.message || 'unknown error')
    process.exit(1)
  }

  spinner.succeed(`"${realDeckName}"含卡片${cards.length}张: ${targetFile}`)
}
