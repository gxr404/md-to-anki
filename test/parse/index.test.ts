import * as fs from 'node:fs/promises'
import axios from 'axios'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { parseCard, parseCardList, parseMarkdown, splitByCards } from '../../src/parse'
import { getConfig, initConfig } from '../../src/config'

vi.mock('node:fs/promises')

const normalMdStr = `
# Test Title

## test

test1

\`\`\`js
var a = 13
\`\`\`

> test

**abc**13123

## test2

## test3

132

## test4

![](./1.jpg)

## test5

![](https://xx.com/1.jpg)

> test

[#tag1]
[#tag2]
[#tag3]
`

const commentsMdStr = `
# Test Title

## test

test1

\`\`\`js
var a = 13
\`\`\`

> test

**abc**13123

## test2

## test3

132

## test4

![](./1.jpg)

## test5

![](https://xx.com/1.jpg)

> test

[#tag1]
[#tag2]
[#tag3]

<!-- ## test6

test -->
`

beforeEach(() => {
  initConfig({})
})
afterEach(() => {
  vi.restoreAllMocks()
})

describe('parse markdown', () => {
  it('normal parse markdown', async () => {
    vi.spyOn(fs, 'readFile').mockImplementation(async () => normalMdStr)
    vi.spyOn(axios, 'get').mockImplementation(async () => ({
      data: 'remote data',
    }))
    const data = await parseMarkdown('xxx.md')
    expect(data.cards.length).toBe(4)
    expect(data.media.length).toBe(2)
  })
  it ('parse has comments markdown', async () => {
    vi.spyOn(fs, 'readFile').mockImplementation(async () => commentsMdStr)
    vi.spyOn(axios, 'get').mockImplementation(async () => ({
      data: 'remote data',
    }))
    const data = await parseMarkdown('xxx.md')
    expect(data.cards.length).toBe(4)
    expect(data.media.length).toBe(2)
  })
})

describe('split Cards', () => {
  it('normal split cards', async () => {
    const res = await splitByCards(normalMdStr)
    // 此时的rawCards 还不是规范的card列表 可能含 卡片为空的情况
    expect(res.resolveCards.length).toBe(6)
    expect(res.deckName).toBe('Test Title')
    expect(res).toMatchSnapshot()
  })

  it('如果在 markdown 中没有指定卡牌名称，则以默认卡牌名称返回', async () => {
    const res = await splitByCards('## card1\n test')
    expect(res.deckName).toBe(getConfig().deck.defaultName)
  })

  it('正确的卡牌名', async () => {
    const mdStr = `# Title \n test \n test`
    const { deckName } = await splitByCards(mdStr)
    expect(deckName).toBe('Title')
  })

  it('多级卡片名需带上父级名称', async () => {
    const mdStr = `# Title \n test \n## Title2\n test\n### Title3 \n test`
    const { resolveCards } = await splitByCards(mdStr)
    expect(resolveCards).toMatchObject([
      { content: '# Title \n test', levelTitle: [] },
      { content: '## Title2\n test', levelTitle: ['Title2'] },
      {
        content: '### Title2_Title3\n test',
        levelTitle: ['Title2', 'Title3'],
      },
    ])
  })
})

describe('parse card list', () => {
  it('normal parse card list', async () => {
    vi.spyOn(fs, 'readFile').mockImplementation(async () => 'local data')
    vi.spyOn(axios, 'get').mockImplementation(async () => ({
      data: 'remote data',
    }))

    const { resolveCards } = await splitByCards(normalMdStr)
    const res = await parseCardList(resolveCards, 'source.md')
    // 真实的卡片列表
    expect(res.cards.length).toBe(4)
    expect(res.media.length).toBe(2)
    expect(res).toMatchSnapshot()
  })

  it('媒体文件保持唯一', async () => {
    vi.spyOn(fs, 'readFile').mockImplementation(async () => 'local data')
    const mdStr = `## Title\nbody\n![media](./1.jpg)![media](./1.jpg)`

    const { resolveCards } = await splitByCards(mdStr)
    const { media } = await parseCardList(resolveCards, 'test.md')
    expect(media.length).toBe(1)
    expect(media[0].data).toBe('local data')
  })
})

describe('parse card', () => {
  const markdown = '## Title\nbody\n[#tag]()'
  const markdownWithMultipleLines = '## Title\nfront\n%\nback\n[#tag]()'

  it('returns null for a blank string', async () => {
    const data = await parseCard(' ')
    expect(data).toBe(null)
  })
  it('returns null when undefined is passed', async () => {
    const data = await parseCard(undefined)
    expect(data).toBe(null)
  })
  it('creates a card with HTML sides', async () => {
    const card = await parseCard(markdown)
    // console.log(card)
    expect(card).toBeTruthy()
    expect(card!.front).toEqual('<h2>Title</h2>\n')
    expect(card!.back).toEqual('<p>body</p>\n')
    expect(card!.tags.length).toEqual(1)
    expect(card!.tags[0]).toEqual('tag')
  })
  it('creates a multi-line card with HTML sides', async () => {
    const card = await parseCard(markdownWithMultipleLines)
    expect(card).toBeTruthy()
    expect(card!.front.replace(/\n/g, '')).toEqual('<h2>Title</h2><p>front</p>')
    expect(card!.back).toEqual('<p>back</p>\n')
    expect(card!.tags.length).toEqual(1)
    expect(card!.tags[0]).toEqual('tag')
  })
  it('skips first blank lines for back', async () => {
    const card = await parseCard('## Title\n \nbody')
    expect(card).toBeTruthy()
    expect(card!.front.replace(/\n/g, '')).toEqual('<h2>Title</h2>')
    expect(card!.back.replace(/\n/g, '')).toEqual('<p>body</p>')
  })
})
