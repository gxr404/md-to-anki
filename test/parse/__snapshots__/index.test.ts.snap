// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse card list > normal parse card list 1`] = `
{
  "cards": [
    {
      "back": "<p>test1</p>
<pre><code class="core lang-js"><pre class="shiki vitesse-dark" style="background-color:#121212;color:#dbd7caee" tabindex="0"><code><span class="line"><span style="color:#CB7676">var</span><span style="color:#BD976A"> a</span><span style="color:#666666"> =</span><span style="color:#4C9A91"> 13</span></span></code></pre>
</code></pre><blockquote>
<p>test</p>
</blockquote>
<p><strong>abc</strong>13123</p>
",
      "front": "<h2>test</h2>
",
      "tags": [],
    },
    {
      "back": "<p>132</p>
",
      "front": "<h2>test3</h2>
",
      "tags": [],
    },
    {
      "back": "<p><img src="40e4a2515baa0cd7cc47f7ce048b6426.jpg" alt=""></p>
",
      "front": "<h2>test4</h2>
",
      "tags": [],
    },
    {
      "back": "<p><img src="39ee43d362145cff5c1cea14f0f39840.jpg" alt=""></p>
<blockquote>
<p>test</p>
</blockquote>
",
      "front": "<h2>test5</h2>
",
      "tags": [
        "tag1",
        "tag2",
        "tag3",
      ],
    },
  ],
  "media": [
    {
      "checksum": "40e4a2515baa0cd7cc47f7ce048b6426",
      "data": "local data",
      "fileName": "40e4a2515baa0cd7cc47f7ce048b6426.jpg",
      "srcStr": "src="40e4a2515baa0cd7cc47f7ce048b6426.jpg"",
    },
    {
      "checksum": "39ee43d362145cff5c1cea14f0f39840",
      "data": "remote data",
      "fileName": "39ee43d362145cff5c1cea14f0f39840.jpg",
      "srcStr": "src="39ee43d362145cff5c1cea14f0f39840.jpg"",
    },
  ],
}
`;

exports[`parse card list > normal split cards 1`] = `
{
  "cards": [
    {
      "back": "<p>test1</p>
<pre><code class="core lang-js"><pre class="shiki vitesse-dark" style="background-color:#121212;color:#dbd7caee" tabindex="0"><code><span class="line"><span style="color:#CB7676">var</span><span style="color:#BD976A"> a</span><span style="color:#666666"> =</span><span style="color:#4C9A91"> 13</span></span></code></pre>
</code></pre><blockquote>
<p>test</p>
</blockquote>
<p><strong>abc</strong>13123</p>
",
      "front": "<h2>test</h2>
",
      "tags": [],
    },
    {
      "back": "<p>132</p>
",
      "front": "<h2>test3</h2>
",
      "tags": [],
    },
    {
      "back": "<p><img src="40e4a2515baa0cd7cc47f7ce048b6426.jpg" alt=""></p>
",
      "front": "<h2>test4</h2>
",
      "tags": [],
    },
    {
      "back": "<p><img src="39ee43d362145cff5c1cea14f0f39840.jpg" alt=""></p>
<blockquote>
<p>test</p>
</blockquote>
",
      "front": "<h2>test5</h2>
",
      "tags": [
        "tag1",
        "tag2",
        "tag3",
      ],
    },
  ],
  "media": [
    {
      "checksum": "40e4a2515baa0cd7cc47f7ce048b6426",
      "data": "local data",
      "fileName": "40e4a2515baa0cd7cc47f7ce048b6426.jpg",
      "srcStr": "src="40e4a2515baa0cd7cc47f7ce048b6426.jpg"",
    },
    {
      "checksum": "39ee43d362145cff5c1cea14f0f39840",
      "data": "remote data",
      "fileName": "39ee43d362145cff5c1cea14f0f39840.jpg",
      "srcStr": "src="39ee43d362145cff5c1cea14f0f39840.jpg"",
    },
  ],
}
`;

exports[`split Cards > normal split cards 1`] = `
{
  "deckName": "Test Title",
  "rawCards": [
    "# Test Title",
    "## test

test1

\`\`\`js
var a = 13
\`\`\`

> test

**abc**13123",
    "## test2",
    "## test3

132",
    "## test4

![](./1.jpg)",
    "## test5

![](https://xx.com/1.jpg)

> test

[#tag1]
[#tag2]
[#tag3]",
  ],
}
`;
